### PR TITLE
Add "filters" option to support removing sensitive properties from payload

### PR DESCRIPTION
**Pull request by bengourley**
### Why?

This is an existing feature of most notifiers, including bugsnag-node, but was not yet implemented in the browser notifier.

### What?

The option can be used like so:

```js
const bugsnagClient = bugsnag({
  apiKey: 'api_key',
  filters: [ 'password', 'access_token', 'refresh_token' ]
})
```

If not supplied, the default value is `[ 'password' ]`. __Should we add anything else to the default list?__

### How?

This feature is implemented as a configurable array of property names. The content of any property in the payload whose key name matches an entry in the array will be replaced with the string `[FILTERED]`.

This implementation – rather than traversing specific parts of the payload – will replace any matching properties anywhere. 

#### Known drawback

This means that if a value is provided such as 'events' or 'stacktrace', then the resulting payload will not be valid for the error reporting api. __Is this a problem?__

The advantage of doing it this way means that we can leverage the built in [`JSON.stringify(obj, REPLACER)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#The_replacer_parameter#The_replacer_parameter) functionality, meaning:
- it's a simpler implementation than writing our own
- it's more performant (using the serialisation traversal to do the filtering means the data structure is only walked once)
- it results smaller output for the browser bundle

### CODE REVIEW

**Comment by martin308** [comment]
We did a bunch of discussion around filtering the entire payload when we redid [bugsnag-dotnet](https://github.com/bugsnag/bugsnag-dotnet). The conclusion was that it should only filter from `user`, `app`, `metaData`, `breadcrumbs`, `request`, and `device` sections of the payload. I'll see if I can dig up any of the conversation from the time

File: packages/core/config.js [hunk 1]
```diff
@@ -90,6 +90,11 @@ module.exports.schema = {
         (accum, method) => accum && typeof value[method] === 'function',
         true
       ))
+  },
+  filters: {
+    defaultValue: () => [ 'password' ],
+    message: 'should be an array of strings',
+    validate: value => isArray(value) && value.length === filter(value, s => typeof s === 'string').length
```
Comment (fractalwrench) [inline packages/core/config.js]: it'd be nice to cover this validation with a couple of unit tests if that hasn't already been done elsewhere.

**Comment by loopj** [comment]
Since the worst case of a customer adding "stacktrace" or "events" to their filters is pretty catastrophic, and will likely cause a ton of confusion and wasted time for customer and support, I'd prefer if we stuck with the notifier library specification on this one and just filter user, app, metaData, breadcrumbs, request, and device.

You could potentially still use a payload-level JSON replacer function, as long as you mark the objects we wish to filter. Eg:

```javascript
// Mark the object
this.app = {
  _filter: true
}
```

**Comment by loopj** [comment]
Another quick question - will this support "string contains" as well as "regex matches" filtering? That would be a nice plus, and AFAIK would be pretty easy to implement in JS. We have this on some other platforms but is not required by the notifier library spec.

**Comment by bengourley** [comment]
Ok – based on your feedback @loopj I've pushed up a different implementation. It now accepts a list of strings/regexes and they are only apploed to the desired portions of the payload. Most of the heavy lifting is done in @bugsnag/safe-json-stringify as I still wanted to avoid additional traversals if at all possible.

Check out https://github.com/bugsnag/safe-json-stringify/pull/1.

File: packages/delivery-node/test/delivery.test.js [hunk 2]
```diff
@@ -32,7 +32,8 @@ describe('delivery:node', () => {
       const payload = { sample: 'payload' }
       const config = {
         apiKey: 'aaaaaaaa',
-        endpoints: { notify: `http://0.0.0.0:${server.address().port}/notify/` }
+        endpoints: { notify: `http://0.0.0.0:${server.address().port}/notify/` },
+        filters: []
```
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/delivery-node/test/delivery.test.js]: these unit tests don't go via the full config schema and so don't get any default values, so every part of config that they depend on needs to be provided

File: packages/browser/features/filters.feature [hunk 3]
```diff
@@ -0,0 +1,30 @@
+@filters
+Feature: Filtering sensitive content from payload
```
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/browser/features/filters.feature]: The properties are already fairly nested in the payload.

File: packages/core/package.json [hunk 4]
```diff
@@ -19,7 +19,7 @@
   "license": "MIT",
   "dependencies": {
     "@bugsnag/cuid": "^3.0.0",
-    "@bugsnag/safe-json-stringify": "^2.1.0",
+    "@bugsnag/safe-json-stringify": "3.0.0-0",
```
Comment (bengourley) [inline packages/core/package.json]: once the PR on @bugsnag/safe-json-stringify is merged and published we should update to a non prerelease version

File: packages/core/config.js [hunk 5]
```diff
@@ -90,6 +90,14 @@ module.exports.schema = {
         (accum, method) => accum && typeof value[method] === 'function',
         true
       ))
+  },
+  filters: {
+    defaultValue: () => [ 'password' ],
```
Comment (bengourley) [inline packages/core/config.js]: should we add any others to this array by default?

**Comment by bengourley** [comment]
> You could potentially still use a payload-level JSON replacer function, as long as you mark the objects we wish to filter.

It's not possible to do this inside a replacer function – in each call of the replacer, the only context you have is the direct parent of the property you are looking at, not the entire "path".

File: packages/core/config.js (hunk 5 above)
Comment (martin308) [reply packages/core/config.js]: I think keeping it minimal by default is good, so I'd leave it as this

File: packages/core/types/common.d.ts [hunk 6]
```diff
@@ -16,6 +16,7 @@ export interface IConfig {
   user?: object | null;
   metaData?: object | null;
   logger?: ILogger | null;
+  filters?: string[];
```
Comment (martin308) [inline packages/core/types/common.d.ts]: Should this be `filters?: (string|RegExp)[];`?

Comment (bengourley) [reply packages/core/types/common.d.ts]: Yeah, or it needs to allow a mixture which I'm not sure that does? I'll fix it, thanks!

Comment (bengourley) [reply packages/core/types/common.d.ts]: your example was correct

File: packages/core/lib/json-payload.js [hunk 7]
```diff
@@ -1,7 +1,20 @@
...
+  'events.[].metaData',
+  'events.[].user',
+  'events.[].breadcrumbs',
+  'events.[].request',
+  'events.[].device',
+  // session
+  'device',
+  'app',
+  'user'
+]
```
Comment (martin308) [inline packages/core/lib/json-payload.js]: should these just be two separate arrays? We have the two separate methods that use this array and it seems like there is no sharing between the two sets?

File: packages/core/lib/json-payload.js [hunk 8]
```diff
@@ -10,8 +23,12 @@ module.exports = report => {
...
     }
-    payload = jsonStringify(report)
+    payload = jsonStringify(report, null, null, { filterPaths, filterKeys })
     if (payload.length > 10e5) throw new Error('payload exceeded 1MB limit')
   }
   return payload
 }
+
+module.exports.session = (report, filterKeys) => {
+  return jsonStringify(report, null, null, { filterPaths, filterKeys })
```
Comment (martin308) [inline packages/core/lib/json-payload.js]: should we check the payload size here as well?

File: packages/browser/bin/minify [hunk 9]
```diff
@@ -4,5 +4,5 @@ cat - | ./node_modules/.bin/uglifyjs \
   --compress \
   --mangle \
   --ie8 \
-  --source-map "content=inline,url=$(basename $1).map" \
+  --source-map "includeSources,content=inline,url=$(basename $1).map" \
```
[the remark this answers is not in the document]
Comment (bengourley) [reply packages/browser/bin/minify]: No but I noticed it was wrong when trying to debug this changeset in development.

**Pull request merged**